node/core/dispute-coordinator/src/initialized.rs

```diff
@@ -779,6 +800,12 @@ impl Initialized {
 			"Number of validators"
 		);
 
+		// If we have already fetched the votes - use them here.
+		let votes_in_db = match maybe_votes_in_db {
```

Comment:
Ok, this is super confusing. What was wrong with the [old code](https://github.com/paritytech/polkadot/blob/97bd688932499e309c085f67298ad3deb98b7420/node/core/dispute-coordinator/src/initialized.rs#L760) where we always fetched the votes above already?

Reply:
This is what I meant [here](https://github.com/paritytech/polkadot/pull/6968#discussion_r1155891129) :)

Before doing anything in `handle_import_statements`, I need the SessionInfo. It might be in `MaybeCandidateReceipt` or it might not. If it is not - I query the db.

But later when the `CandidateVoteState` the logic is reversed. Query the db, if it is empty - try `MaybeCandidateReceipt` and if there is no `CandidateReceipt` there too - fail.

To avoid two db lookups I've added `maybe_votes_in_db`, which is an option. If it is Some(_) - the result from the database query is available and can be used. If not - do the necessary query while generating `CandidateVoteState`.

It's ugly and complicated - I don't argue at all. So I'd prefer to keep the old approach where the `CandidateReceipt` was obtained from `ActiveLeavesUpdate` and passed down. Unless there is an invariant I don't see which allows the code to be simplified.

Reply:
> Before doing anything in `handle_import_statements`, I need the SessionInfo. It might be in `MaybeCandidateReceipt` or it might not. If it is not - I query the db.

But why not always query the db there already?

Reply:
I got a bit carried away. I was trying to avoid redundant checks, but the end result was way worse compared to leaving them:)

It should be okay now.